From c31d8e6684096525df651aeb826f6874a75c5b76 Mon Sep 17 00:00:00 2001
From: Don Zickus <dzickus@redhat.com>
Date: Fri, 12 Sep 2014 18:05:30 -0400
Subject: [powerpc] Emulate sync instruction variants

Message-id: <1410545655-205645-102-git-send-email-dzickus@redhat.com>
Patchwork-id: 94125
O-Subject: [RHEL7 PATCH 101/626] powerpc: Emulate sync instruction variants
Bugzilla: 1127366
RH-Acked-by: Prarit Bhargava <prarit@redhat.com>
RH-Acked-by: Steve Best <sbest@redhat.com>
RH-Acked-by: Tony Camuso <tcamuso@redhat.com>
RH-Acked-by: David Gibson <dgibson@redhat.com>

https://bugzilla.redhat.com/show_bug.cgi?id=1127366

commit 9863c28a2af90a56c088f5f6288d7f6d2c923c14
Author: James Yang <James.Yang@freescale.com>
Date:   Wed Jul 3 16:26:47 2013 -0500

    powerpc: Emulate sync instruction variants

    Reserved fields of the sync instruction have been used for other
    instructions (e.g. lwsync).  On processors that do not support variants
    of the sync instruction, emulate it by executing a sync to subsume the
    effect of the intended instruction.

    Signed-off-by: James Yang <James.Yang@freescale.com>
    [scottwood@freescale.com: whitespace and subject line fix]
    Signed-off-by: Scott Wood <scottwood@freescale.com>
---

Signed-off-by: Jarod Wilson <jarod@redhat.com>

diff --git a/arch/powerpc/include/asm/ppc-opcode.h b/arch/powerpc/include/asm/ppc-opcode.h
index 98a0851..99f8790 100644
--- a/arch/powerpc/include/asm/ppc-opcode.h
+++ b/arch/powerpc/include/asm/ppc-opcode.h
@@ -143,6 +143,8 @@
 #define PPC_INST_LSWX   0x7c00042a
 #define PPC_INST_LWARX   0x7c000028
 #define PPC_INST_LWSYNC   0x7c2004ac
+#define PPC_INST_SYNC   0x7c0004ac
+#define PPC_INST_SYNC_MASK  0xfc0007fe
 #define PPC_INST_LXVD2X   0x7c000698
 #define PPC_INST_MCRXR   0x7c000400
 #define PPC_INST_MCRXR_MASK  0xfc0007fe
diff --git a/arch/powerpc/kernel/traps.c b/arch/powerpc/kernel/traps.c
index b2a8277..6fb62e0 100644
--- a/arch/powerpc/kernel/traps.c
+++ b/arch/powerpc/kernel/traps.c
@@ -1015,6 +1015,13 @@ static int emulate_instruction(struct pt_regs *regs)
   return emulate_isel(regs, instword);
  }
 
+ /* Emulate sync instruction variants */
+ if ((instword & PPC_INST_SYNC_MASK) == PPC_INST_SYNC) {
+  PPC_WARN_EMULATED(sync, regs);
+  asm volatile("sync");
+  return 0;
+ }
+
 #ifdef CONFIG_PPC64
  /* Emulate the mfspr rD, DSCR. */
  if ((((instword & PPC_INST_MFSPR_DSCR_USER_MASK) ==
-- 
1.7.1